gr-qtgui/lib/fosphor_formatter_impl.cc

```diff
@@ -12,205 +13,223 @@
-                     gr_vector_const_void_star& input_items,
-                     gr_vector_void_star& output_items) final
-    {
-        const int items_to_process = std::min<int>({ ninput_items[0],
-                                                     ninput_items[1],
-                                                     ninput_items[2],
-                                                     d_histo_decim - d_histo_count });
-        if (items_to_process == 0) {
-            return 0;
+        // Decimate input by ignoring
+        ++d_input_count;
+        ++n;
+        if (d_input_count < d_input_decim) {
+            in += d_fft_size;
+            continue;
         }
-        // We are guaranteed to have space for at least one histogram output,
-        // because we called set_output_multiple() in the ctor on the histo proc.
+        d_input_count = 0;
 
-        const unsigned char* in_logfft_b =
-            static_cast<const unsigned char*>(input_items[0]);
-        const float* in_logfft_f = static_cast<const float*>(input_items[1]);
-        const float* in_logfft_avg = static_cast<const float*>(input_items[2]);
-        unsigned char* out = static_cast<unsigned char*>(output_items[0]);
+        gr_complex* fft_in = d_fft->get_inbuf();
+        volk_32fc_32f_multiply_32fc(fft_in, in, d_window.data(), d_fft_size);
+        d_fft->execute();
+        gr_complex* fft_out = d_fft->get_outbuf();
+        const float scale_factor =
+            1.0f * 1.0f / static_cast<float>(d_fft_size * d_fft_size);
+        // values are divided by normalizationFactor before power calc, so invert
+        volk_32fc_s32f_power_spectrum_32f(d_logfft.data() + d_fft_size / 2,
+                                          fft_out,
+                                          1.0f / scale_factor,
+                                          d_fft_size / 2);
+        volk_32fc_s32f_power_spectrum_32f(d_logfft.data(),
+                                          fft_out + d_fft_size / 2,
+                                          1.0f / scale_factor,
+                                          d_fft_size / 2);
+        volk_32f_s32f_add_32f(d_logfft.data(), d_logfft.data(), 100.0f, d_fft_size);
 
         // Decay previous max hold value
         volk_32f_s32f_multiply_32f(
             d_maxhold_buf.data(), d_maxhold_buf.data(), d_epsilon, d_fft_size);
-        for (int i; i < items_to_process; i++) {
-            //// Update max hold:
-            // Compare with current max
-            volk_32f_x2_max_32f(
-                d_maxhold_buf.data(), d_maxhold_buf.data(), in_logfft_f, d_fft_size);
-            //// Update hit counter
+
+        // Update max
+        volk_32f_x2_max_32f(
+            d_maxhold_buf.data(), d_maxhold_buf.data(), d_logfft.data(), d_fft_size);
+
+        ++d_histo_count;
+        if (d_histo_count == d_histo_decim) {
+
+            ++n_histo;
+            d_histo_count = 0;
+
+            volk_32f_s32f_x2_convert_8u(
```

Comment:
Is this kernel new? I remember looking for something like this and blanking.

Reply:
In fact, `volk` `master` does not have this kernel.

Reply:
VOLK uses `main`

Reply:
This is new since v3.0.0, so I guess I can't use it here and will have to use `volk_32f_s32f_convert_8i` or something.